File: frontend/src/captions/lang.ts
/**
 * Which subtitle track a language tag belongs to, and how to order a line's two
 * languages.
 *
 * Mirrors `Lang.java` on the server, which does the same reduction before
 * deciding what to translate into. Kept in step by hand, like the signaling
 * envelope: both sides have to agree that `cmn-Hant-TW` is Chinese, or the
 * server translates a line the browser then files under the wrong heading.
 */

export type Track = 'zh' | 'en'

/**
 * The track a BCP-47 tag belongs to, or null when it is neither.
 *
 * Matched on the primary subtag — except that Mandarin and Cantonese carry
 * ISO-639-3 primary subtags (`cmn`, `yue`) that do not begin with `zh` at all,
 * which is the case a naive `startsWith('zh')` gets wrong on the exact locale a
 * Taiwanese room speaks.
 */
export function track(tag: string | undefined): Track | null {
  if (!tag) return null
  const primary = tag.trim().toLowerCase().split(/[-_]/)[0]
  if (['zh', 'cmn', 'yue', 'nan', 'hak', 'wuu'].includes(primary)) return 'zh'
  if (primary === 'en') return 'en'
  return null
}

export interface Utterance {
  text: string
  lang: string
  translation?: string
  translationLang?: string
}

export interface Rendered {
  text: string
  lang: string
  /** True for the words that were actually said, false for the translation. */
  original: boolean
}

/**
 * A line's languages, in a fixed order: Chinese first, then English.
 *
 * <p><strong>Both are always shown, and nobody chooses.</strong> An earlier
 * version put the reader's picked language on top, which meant the subtitle
 * changed shape depending on who was reading and made the language selector
 * mean two unrelated things at once — what I am speaking, and what I want to
 * read. In a room that is half Chinese and half English, everybody needs both
 * lines anyway; what they do not need is for the pair to swap places between one
 * utterance and the next depending on which language it was spoken in.
 *
 * <p>A fixed order also means the eye learns where to look. Ordering by
 * "original first" would put Chinese on top for one sentence and English on top
 * for the next, in the same conversation.
 *
 * <p>Untranslated lines, and lines in neither language, return the single row
 * they have — never a blank second line implying a translation that is not there.
 */
export function bilingual(line: Utterance): Rendered[] {
  const rows: Rendered[] = [{ text: line.text, lang: line.lang, original: true }]
  if (line.translation) {
    rows.push({
      text: line.translation,
      lang: line.translationLang ?? '',
      original: false,
    })
  }
  if (rows.length < 2) return rows

  const rank = (row: Rendered) => {
    const t = track(row.lang)
    // Anything unrecognised sorts last rather than jumping the queue: it is the
    // row we know least about.
    return t === 'zh' ? 0 : t === 'en' ? 1 : 2
  }
  return [...rows].sort((a, b) => rank(a) - rank(b))
}

File: backend/src/main/java/com/warroomlive/transcript/Lang.java
package com.warroomlive.transcript;

import java.util.Locale;
import java.util.Optional;

/**
 * The two languages this room translates between, and how to recognise them in
 * whatever tag a browser hands over.
 *
 * <p>Speech recognition reports BCP-47, and it is not tidy: Chrome says
 * {@code cmn-Hant-TW} for Taiwanese Mandarin, {@code zh-TW} in older builds,
 * {@code yue-Hant-HK} for Cantonese, {@code en-US} and {@code en-GB} for
 * English. All the Chinese varieties are one subtitle track here and all the
 * English ones are the other, so the tag is reduced to {@code zh} or {@code en}
 * and everything else is declined rather than guessed at.
 *
 * <p>Declining is deliberate. Somebody speaking Japanese into a room configured
 * for Chinese and English should see no translation at all, rather than a
 * confident mistranslation into a language they did not ask for.
 */
public final class Lang {

    public static final String ZH = "zh";
    public static final String EN = "en";

    private Lang() {
    }

    /**
     * The subtitle track a BCP-47 tag belongs to, or empty when it is neither.
     *
     * <p>Matched on the primary subtag, except that Mandarin and Cantonese carry
     * ISO-639-3 primary subtags ({@code cmn}, {@code yue}) that do not begin
     * with {@code zh} at all — the case a naive {@code startsWith("zh")} gets
     * wrong on the exact locale a Taiwanese room uses.
     */
    public static Optional<String> track(String tag) {
        if (tag == null || tag.isBlank()) return Optional.empty();
        String primary = tag.trim().toLowerCase(Locale.ROOT).split("[-_]")[0];
        return switch (primary) {
            // zh: the macrolanguage. cmn/yue/nan/hak/wuu: the varieties Chrome
            // actually reports for Mandarin, Cantonese, Hokkien, Hakka and Wu.
            case "zh", "cmn", "yue", "nan", "hak", "wuu" -> Optional.of(ZH);
            case "en" -> Optional.of(EN);
            default -> Optional.empty();
        };
    }

    /** The other track — what a line gets translated into. Empty when the tag is neither. */
    public static Optional<String> counterpart(String tag) {
        return track(tag).map(t -> ZH.equals(t) ? EN : ZH);
    }

    /** How the model is asked for the target, in words rather than a code. */
    public static String describe(String track) {
        return ZH.equals(track) ? "Traditional Chinese (繁體中文)" : "English";
    }
}
